File: circular-queue.c
#include <stdio.h>

#define SIZE 5

int items[SIZE];
int front = -1, rear = -1;

// check if the queue is full
int isFull() {
  if ((front == (rear + 1) % SIZE) || (front == 0 && rear == SIZE - 1)){
    return 1;
  }else{
    return 0;
  }
}

// check if the queue is empty
int isEmpty() {
  if (front == -1){
    return 1;
  }else{
    return 0;
  }
}

// adding an element
void enQueue(int element) {
  if (isFull())
    printf("\n Queue is full!! \n");
  else {
    if (front == -1){
        front = 0;
    };
    rear = (rear + 1) % SIZE;
    items[rear] = element;
    printf("\n Inserted -> %d", element);
  }
}

// removing an element
void deQueue() {
  int element;
  if (isEmpty()) {
    printf("\n Queue is empty !! \n");
    return;
  } else {
    element = items[front];

    if (front == rear) {
      front = -1;
      rear = -1;
    }
    else {
      front = (front + 1) % SIZE;
    }

    printf("\n Deleted element -> %d \n", element);
  }
}

// display the queue
void display() {
  int i;
  if (isEmpty())
    printf(" \n Empty Queue\n");
  else {
    printf("\n Front -> %d ", front);
    printf("\n Items -> ");

    i = front;
    do {
      printf("%d ", items[i]);
      i = (i + 1) % SIZE;
    } while (i != (rear + 1) % SIZE);
    
    printf("\n Rear -> %d \n", rear);
  }
}

int main() {
  deQueue();

  enQueue(1);
  enQueue(2);
  enQueue(3);
  enQueue(4);
  enQueue(5);

  enQueue(6);

  display();
  deQueue();

  display();

  enQueue(7);
  display();

  enQueue(8);

  return 0;
}

File: circular-queue.js
class CircularQueue {
  /**
   * Constructor to initialize the circular queue
   * @param {number} size
   */
  constructor(size) {
    this.size = size;
    this.items = new Array(size);
    this.front = -1;
    this.rear = -1;
  }

  /**
   * Check if the queue is full
   * @returns {boolean}
   */
  isFull() {
    return (
      this.front === (this.rear + 1) % this.size ||
      (this.front === 0 && this.rear === this.size - 1)
    );
  }

  /**
   * Check if the queue is empty
   * @returns {boolean}
   */
  isEmpty() {
    return this.front === -1;
  }

  /**
   * Add an element to the queue
   * @param {*} element
   */
  enQueue(element) {
    if (this.isFull()) {
      console.log("\nQueue is full!!");
    } else {
      if (this.front === -1) {
        this.front = 0;
      }
      this.rear = (this.rear + 1) % this.size;
      this.items[this.rear] = element;
      console.log(`\nInserted -> ${element}`);
    }
  }

  /**
   * Remove an element from the queue
   */
  deQueue() {
    if (this.isEmpty()) {
      console.log("\nQueue is empty!!");
    } else {
      let element = this.items[this.front];
      if (this.front === this.rear) {
        this.front = -1;
        this.rear = -1;
      } else {
        this.front = (this.front + 1) % this.size;
      }
      console.log(`\nDeleted element -> ${element}`);
    }
  }

  /**
   * Display the current elements in the queue
   */
  display() {
    if (this.isEmpty()) {
      console.log("\nEmpty Queue");
    } else {
      console.log(`\nFront -> ${this.front}`);
      console.log("Items ->", this.getQueueElements().join(" "));
      console.log(`Rear -> ${this.rear}`);
    }
  }

  /**
   * Helper function to get all elements of the queue in the correct order
   * @returns {Array}
   */
  getQueueElements() {
    let result = [];
    let i = this.front;
    do {
      result.push(this.items[i]);
      i = (i + 1) % this.size;
    } while (i !== (this.rear + 1) % this.size);
    return result;
  }
}

/**
 * Example Usage
 */
const cq = new CircularQueue(5);

cq.deQueue();

cq.enQueue(1);
cq.enQueue(2);
cq.enQueue(3);
cq.enQueue(4);
cq.enQueue(5);

cq.enQueue(6);

cq.display();

cq.deQueue();
cq.display();

cq.enQueue(7);
cq.display();

cq.enQueue(8);
